Include tash in the Task bundle build

zregvart opened the pull request:
This includes the tool to generate the Trusted Artifacts Task variants in the `hack/build-and-push.sh` script, now if the task has a `recipe.yaml` file within its directory the tool will be used to regenerate the Trusted Artifact Task from the base Task.

Fixes https://issues.redhat.com/browse/EC-650

lcarva commented on task/buildah-oci-ta/0.1/buildah-oci-ta.yaml:
```diff
@@ -16,87 +16,96 @@ spec:
+      central.
+    name: JAVA_COMMUNITY_DEPENDENCIES
   stepTemplate:
+    computeResources: {}
```
NIT: It would be nice to just omit this attribute.

lcarva commented on task/buildah-oci-ta/0.1/buildah-oci-ta.yaml:
```diff
@@ -1,13 +1,13 @@
+  annotations:
+    tekton.dev/pipelines.minVersion: 0.12.1
+    tekton.dev/tags: image-build, appstudio, hacbs
+  creationTimestamp: null
```
NIT: It would be nice to omit this.

zregvart replied:
Yeah I was a bit puzzled why it is there, but I didn't really want to address it, I guess it might be missing `omitempty`? I'll look into it...

zregvart replied:
The issue with timestamps is that they're using [zero value](https://pkg.go.dev/time#Time.IsZero) of `time.Time`, and the custom Marshaller implementation converts that value to `null`, there is no option to omit `null` values in golang's json. New version of tkn-fmt handles this

lcarva commented on task/buildah-oci-ta/0.1/buildah-oci-ta.yaml:
```diff
@@ -183,7 +193,7 @@ spec:
 
       # Setting new namespace to run buildah - 2^32-2
-      echo 'root:1:4294967294' | tee -a /etc/subuid >> /etc/subgid
+      echo 'root:1:4294967294' | tee -a /etc/subuid >>/etc/subgid
```
NIT: I do prefer the whitespace, but it's not a big deal if it's much of an effort to keep those in.

zregvart replied:
Done by the shell parser/formatter and the https://pkg.go.dev/mvdan.cc/sh/v3@v3.8.0/syntax#SpaceRedirects doesn't help it seems

lcarva commented on task/buildah-oci-ta/0.1/buildah-oci-ta.yaml:
```diff
@@ -274,29 +284,25 @@ spec:
-  - name: sbom-syft-generate
+  - computeResources: {}
     image: quay.io/redhat-appstudio/syft:v0.105.1@sha256:1910b829997650c696881e5fc2fc654ddf3184c27edb1b2024e9cb2ba51ac431
-    # Respect Syft configuration if the user has it in the root of their repository
```
NIT: Any chance we can keep the comment?

zregvart replied:
The issue here is that the YAML is unmarshalled to a golang struct and comments are lost, I missed this because the test code also does this, this might take some tinkering

zregvart replied:
I had a try of restoring comments, the issue is that comments that are part of sequences loose their position as the tool adds or removes elements from the sequence, i.e. a comment on `spec.steps[n]` can't be guaranteed to be placed on the same place, and the option of tracking the comments by each item's identifier (i.e. step name) seemed like a lot of work. The attempt I had added comments to wrong places and for whatever reason (read: implementation bug on my side) did not restore this comment in particular. So the lazy option is to loose the comments, at least for now.

lcarva reviewed (approved):
This is great!

zregvart commented on appstudio-utils/Dockerfile:
```diff
@@ -8,6 +8,7 @@ RUN curl -L https://github.com/tektoncd/cli/releases/download/v0.32.2/tkn_0.32.2
 RUN curl -L https://github.com/sigstore/rekor/releases/download/v0.5.0/rekor-cli-linux-amd64 -o /usr/bin/rekor-cli && chmod +x /usr/bin/rekor-cli
 RUN curl -L https://github.com/open-policy-agent/conftest/releases/download/v0.32.0/conftest_0.32.0_Linux_x86_64.tar.gz | tar -xz --no-same-owner -C /usr/bin
 RUN curl -L https://github.com/enterprise-contract/ec-cli/releases/download/snapshot/ec_linux_amd64 -o /usr/bin/ec && chmod +x /usr/bin/ec && ec version
+ADD --chmod=755 --chown=0:0 https://github.com/enterprise-contract/hacks/releases/download/latest/tash /usr/local/bin
```
Separate repository? Easily done, though I was also hoping the tool will not be that long lived...

zregvart replied:
Decided to be lazy and work with hacks

zregvart commented on hack/build-and-push.sh:
```diff
@@ -7,6 +7,20 @@ QUAY_ORG=redhat-appstudio-tekton-catalog
 WORKDIR=$(mktemp -d --suffix "-$(basename "${BASH_SOURCE[0]}" .sh)")
 
+if ! command -v tash &> /dev/null; then
+  echo INFO: tash command is not available will download and use the latest version
```
I agree, I did not want to end up troubleshooting and assisting folk in getting that latest version including the fix. And I'm sure there will be fixes as we progress here. I also considered moving the code here and using `go run` but I also use go 1.22 to build so that might be an issue for some folk (not sure?). I could expand this to prompt if the user wants the tool to be downloaded/installed and include a digest or similar.
This will be run only locally, the CI has the latest version included in the appstudio-util image...

zregvart replied:
One small affordance I did is to do a sha256sum over the downloaded file...

zregvart commented on task/buildah-oci-ta/0.1/buildah-oci-ta.yaml:
```diff
@@ -1,13 +1,13 @@
----
```
I should upload my wonderful tkn fmt tool, I was making some refinements and I think I broke something. With it this would be `diff -u <(tkn fmt <(curl -sL $MAIN)) <(tkn fmt <(curl -sL $PR))` or something along those lines...

zregvart replied:
Dropping a link to https://github.com/zregvart/tkn-fmt

zregvart commented:
Thanks for the suggestions I'll look into addressing them, other reviews welcome...

chmeliik commented on hack/build-and-push.sh:
```diff
@@ -134,7 +162,12 @@ do
+done < <(find task/*/*/ -maxdepth 0 -type d | awk -F '/' '{ print $0, $2, $3 }')
+
+if [[ ${errors} -gt 0 ]]; then
+  echo "ERROR: Found differences between base and OCI-TA Task variants, run hack/build-and-push.sh locally and include the regenerated files in your changeset"
```
(this is conceptually the same as the `.github/workflows/check-buildah-remote.yaml` workflow)

zregvart replied:
Yeah, reworked this to run similar to the buildah-remote

chmeliik commented on .tekton/scripts/check-task-pipeline-bundle-repos.sh:
```diff
@@ -22,7 +22,7 @@ echo "Checking existence of task and pipeline bundle repositories ..."
 # tasks
 for task_dir in $(find task/*/*/ -maxdepth 0 -type d); do
-    if [ ! -f $task_dir/kustomization.yaml ]; then
+    if [[ ! -f $task_dir/kustomization.yaml || ! -f $task_dir/recipe.yaml ]]; then
       task_name=$(oc apply --dry-run=client -f "$task_dir/*.yaml" -o jsonpath='{.metadata.name}')
     else
       task_name=$(oc apply --dry-run=client -k "$task_dir" -o jsonpath='{.metadata.name}')
```
Not sure I understand the logic here

```
if the task directory has neither kustomization.yaml nor recipe.yaml:
    read task name directly from task yaml
else
    run kustomize and read task name from the result
```
I think the decision should still be based only on the existence of kustomization.yaml, no?

zregvart replied:
Dropped this logic altogether, not sure if it made sense before, I think it might be leftover from a version I did when the generated files were not versioned...

chmeliik reviewed (commented):
Nice, no more manual syncing of multiple versions of the same task :tada:

zregvart commented:
I think I addressed all the comments, please do have another look. Notable differences from the previous version:
 * TA variant generation is no longer part of `hack/build-and-push.sh` but done via `hack/generate-ta-tasks.sh`, there is now a new workflow to check if the files need to be regenerated
 * the Task YAML files are now tkn-fmt-ed, so diffing base tasks is best done by installing https://github.com/zregvart/tkn-fmt and doing something like `diff <(tkn fmt task.yaml) task-oci-ta.yaml`
 * YAML file comments are difficult to maintain, so the lazy option of ignoring them has been implemented (though this was the same in the previous version)

chmeliik reviewed (commented):
buildah-remote-oci-ta seems to be missing the recipe.yaml, otherwise LGTM :taco:

chmeliik commented:
> buildah-remote-oci-ta seems to be missing the recipe.yaml, otherwise LGTM 🌮

Oh wait, that one is generated from buildah-oci-ta, isn't it

chmeliik commented on hack/generate-ta-tasks.sh:
```diff
@@ -0,0 +1,58 @@
+set -o errexit
+set -o nounset
+set -o pipefail
+set -o posix
```
Just out of curiosity, why do you enable the posix option?

zregvart replied:
I think it was because of the exit code from subshells (or something akin to that), might be required in the previous version when this was included in `hack/build-and-push.sh`.

There is a list of changes in behavior when in the posix mode here: https://www.gnu.org/software/bash/manual/html_node/Bash-POSIX-Mode.html

zregvart commented:
> buildah-remote-oci-ta seems to be missing the recipe.yaml, otherwise LGTM 🌮

It is generated using `hack/generate-buildah-remote.sh` so that's why it is not there.

zregvart commented:
Looks like one of the e2e tests timed out:

```
Summarizing 1 Failure:
  [INTERRUPTED] [build-service-suite Build templates E2E test] HACBS pipelines [It] should eventually finish successfully for component with Git source URL https://github.com/cachito-testing/pip-e2e-test [build, build-templates, HACBS, pipeline, build-templates-e2e, source-build-e2e]
  /github.com/konflux-ci/e2e-tests/tests/build/build_templates.go:207

Ran 24 of 607 Specs in 2327.058 seconds
FAIL! - Interrupted by User -- 23 Passed | 1 Failed | 36 Pending | 547 Skipped
--- FAIL: TestE2E (2327.07s)
FAIL
```

/retest

stuartwdouglas commented on task/buildah-remote-oci-ta/0.1/buildah-remote-oci-ta.yaml:
```diff
@@ -16,43 +16,45 @@ spec:
     When [Java dependency rebuild](https://redhat-appstudio.github.io/docs.stonesoup.io/Documentation/main/cli/proc_enabled_java_dependencies.html) is enabled it triggers rebuilds of Java artifacts.
```
This code is now generated from two different places, as the remote task generation is expecting to generate this from the buildah-oci-ta.

zregvart replied:
From what I understand the `hack/generate-buildah-remote.sh` script will generate `task/buildah-remote-oci-ta/0.1/buildah-remote-oci-ta.yaml`, there is no `recipe.yaml` in `task/buildah-remote-oci-ta/0.1` so tash and the `hack/generate-ta-tasks.sh` script will not try to generate this file.

stuartwdouglas replied:
ok, I missed that it skipped this one, looks good to me.

I have a PR open at https://github.com/konflux-ci/build-definitions/pull/1068 that moves this generator to the build-definitions repo.

zregvart replied:
Yeah I saw that, I think we'll move this from ec/hacks here as well.

lcarva commented:
e2e tests keep failing with:

```
[e2e-tests : e2e-test] • [FAILED] [660.100 seconds]
[e2e-tests : e2e-test] [build-service-suite Build templates E2E test] HACBS pipelines [BeforeAll] triggers PipelineRun for symlink component with source URL https://github.com/redhat-appstudio-qe/devfile-sample-python-basic.git [build, build-templates, HACBS, pipeline, build-templates-e2e, source-build-e2e]
[e2e-tests : e2e-test]   [BeforeAll] /github.com/konflux-ci/e2e-tests/tests/build/build_templates.go:96
[e2e-tests : e2e-test]   [It] /github.com/konflux-ci/e2e-tests/tests/build/build_templates.go:184
[e2e-tests : e2e-test] 
[e2e-tests : e2e-test]   Timeline >>
[e2e-tests : e2e-test]   [FAILED] in [BeforeAll] - /github.com/konflux-ci/e2e-tests/tests/build/build_templates.go:135 @ 06/11/24 15:45:01.939
```

lcarva commented:
I wonder if this needs a rebase now that HAS has been disabled?
